TrackletAlgorithm/VMRouterCM.h

```diff
@@ -0,0 +1,434 @@
+inline T createVMStub(const InputStub<InType> inputStub,
+		const int index, const bool negDisk, const int lutTable[],
+		const int phiCorrTable[], int& slot) {
+
+	// The stub that is going to be returned
+	T stub;
+
+	// Values from InputStub
+	auto z = inputStub.getZ();
+	auto r = inputStub.getR();
+	auto bend = inputStub.getBend();
+	auto phi = inputStub.getPhi();
+	auto phicorr = getPhiCorr<InType>(phi, r, bend, phiCorrTable); // Corrected phi, i.e. phi at nominal radius
+
+	int nbitsr = r.length(); // Number of bits for r
+	int nbitsz = z.length(); // Number of bits for z
+	int nbitsfinerz = stub.getFineZ().length(); // Number of bits for finer/z
+	int nbitsfinephi = stub.getFinePhi().length();  // Number of bits for finephi
+
+	// Number of bits for table indices
+	constexpr int nbitsztable =
+			(Layer) ? nbitsztablelayer : nbitsztabledisk; // Number of MSBs of z used in LUT table
+	constexpr int nbitsrtable =
+			(Layer) ? nbitsrtablelayer : nbitsrtabledisk; // Number of MSBs of r used in LUT table
+	constexpr auto vmbits =
+			(Layer) ? nbitsvmlayer[Layer - 1] : nbitsvmdisk[Disk - 1]; // Number of bits for standard VMs
+	constexpr unsigned int nbitsall = (Layer) ? nbitsallstubs[Layer-1] : nbitsallstubs[N_LAYER+Disk-1]; // Number of bits for the number of Alltub memories in a layer/disk
+
+	// Set values to VMStub
+	stub.setBend(bend);
+	stub.setIndex(index);
+	stub.setFinePhi(
+				iphivmFineBins<InType>(phicorr, nbitsall + vmbits, nbitsfinephi));
+
+	// Indices used to find the rzfine value in LUT table
+	// LUT table returns the top 6 bits of a corrected z
+	// Note: not the index that is being saved to the stub
+	ap_uint<nbitsztable + nbitsrtable> indexz = (z >> (nbitsz - nbitsztable)); // Make z unsigned and take the top "nbitsztable" bits
+	ap_uint<nbitsrtable> indexr = -1;
+
+	constexpr int rbins = (1 << nbitsrtable); // Number of bins in r in LUT table
+
+	if (Disk) {
+		if (negDisk) {
+			indexz = (1 << nbitsztable) -1 - indexz;
+		}
+		indexr = r;
+		if (InType == DISKPS) {
+			indexr = r >> (nbitsr - nbitsrtable); // Take the top "nbitsrtable" bits
+		}
+	} else { // Layer
+		indexr = (r >> (nbitsr- nbitsrtable));// Make r unsigned and take the top "nbitsrtable" bits
+	}
+
+	// The index for LUT table
+	ap_uint<nbitsztable + nbitsrtable> lutIndex = (indexz * rbins) + indexr;
+
+	// Get the LUT value
+	auto lutValue = lutTable[lutIndex];
+
+	// Coarse z. The bin the stub is going to be put in, in the memory
+	int bin = lutValue >> nbitsfinerz; // 3 bits, i.e. max 8 bins within each VM
+
+	if (negDisk) {
+		if (isMEStub) bin += 1 << MEBinsBits; // bin 8-16 are for negative disk
+		else bin += (1 << TEBinsBits)/2; // += 4
+	}
+	
+	auto ivm = phicorr.range(phicorr.length() - nbitsall - 1, phicorr.length() - (nbitsall + vmbits)); //get the phi bits that corresponds to the old vms
+	slot = (Disk && isMEStub) ? ivm * 16 + bin : ivm * 8 + bin; //1 << 3 is the number of bins NBINS?
```

Comment:
Is there a way to write this without the magic 16 and 8? Like instead of `ivm * 16`, you could have something like `(ivm << NBITS)`, where `NBITS` is some named constant equal to 4, and similarly for `ivm * 8`.